### ISSUE TITLE
[AMD only] Minor rendering issues in MGS2&3 (that are not present with nvidia)

### ISSUE THREAD

**Issue by NotBeanz** [RPCS3/rpcs3#6403]
As the title says, MGS2&3 suffer from some exclusive rendering issues on AMD GPUs, that are not present on NVIDIA.

All of these tests were done on the latest MGS4 fixes PR(but the issues are the same on master), with up to date drivers, using Vulkan.
**Firstly:** Snake is still semi-transparent during the tanker chapter.
![2019 08 24-22 32](https://user-images.githubusercontent.com/54376483/63641970-f4d28000-c6bf-11e9-97dd-e3119be4761b.png)
This was captured on my RX 570, but if I switch to my GTX 1050 the issue is resolved. (same base settings were used for each test with deleted shader caches).

**Secondly:** MGS2&3 has random flickering issues on AMD GPUs every now and then. These flickers are not present on NVIDIA hardware.
To illustrate the issue here's a short clip: https://www.youtube.com/watch?v=tv_ZYIpLCxk

**Comment by Snakegodeater** [comment]
Do you have issues on MGS4 on AMD Card ? Can u go ingame with the latest PR ?

**Comment by NotBeanz** [comment]
@Snakegodeater Yes, I can. It is unstable and crashy, but it goes in-game. 
Also in the future keep the discussion on topic.

**Comment by Snakegodeater** [comment]
I have issues too But with R9 380. I am gonna Switch to An Nvidia Card. other Emulators Use OpenGL like Yuzu or Cemu. AMD Sucks There :/. Also RPCS3 Team they Testing on Nvidia Cards too. Ps3 Gpu it's Nvidia too.

**Comment by Asinin3** [comment]
It doesn't matter that the PS3 gpu was Nvidia, and kd-11 the lead graphics dev has both an amd and nvidia gpu and he switches between them a lot and fixes issues on both sides when they appear. We also have Vulkan support so no problems their either. Anyway this is completely off-topic and shouldn't be discussed on GitHub as this is not the place for discussion. Use our Discord server instead: https://discord.me/rpcs3

**Comment by kd-11** [comment]
Missing replay data. No RSX captures or Renderdoc traces provided.

**Comment by NotBeanz** [comment]
@kd-11 The RSX captures didn't work(they always showed a black image), so I used Renderdoc instead.
Here's the link for the semi-transparent Snake RDC on my RX 570: https://drive.google.com/open?id=1USCV0U1RNAK282pdub5D-ZVPpRfIiGyQ

The second issue (the random flickering) is really hard to capture. Should I keep trying or is it understandable from the video that I've provided?

**Comment by kd-11** [comment]
1. Its unlikely you'll be able to capture flickering.
2. When replay is black it means the game uses a different method of handling flip requests, I can still use it since I can see the replays in analysis tools so its usually fine. I'll fix the capture system at a later time.
3. Renderdoc may (hopefully) be enough to identify and fix the bugs here.
4. Since this relies on the MGS4 PR being completed first, I may choose to wait until that is completed before starting work on this. This is because there are about 3 other related PRs lined up.

**Comment by NotBeanz** [comment]
Thanks for clarifying all that.
Now that I know that you can use those RSX captures. I'll provide a link to the semi-transparent Snake RSX capture as well: https://drive.google.com/open?id=1rB08PESIBHfojFJ8BS9bBgCzhcvi7LsK
Also I'd like to mention, that this issue came with PR #6376 

The random flickering on the other hand has existed ever since I've begun testing these games. (but only on AMD GPU's)

**Comment by Snakegodeater** [comment]
@NotBeanz can you open a new issue thread for AMD MGS4 ?

**Comment by NotBeanz** [comment]
@Snakegodeater We have told you this 100 times. MGS4 is unstable for EVERYONE. It doesn't matter what GPU you have. Please stop posting your nonsense everywhere.

**Comment by Snakegodeater** [comment]
Snake looks fine on me. i watched the cutscene when he lands and his body was ok on tanker chapter altough i get the flickering on snake eater

**Comment by NotBeanz** [comment]
Interesting. That might mean that not all AMD cards are affected. Polaris sure is though..

**Comment by Snakegodeater** [comment]
Dunno !!! If it's only polaris RX 4xx or Rx 5xx series ! Did u used the latest driver ?

**Comment by NotBeanz** [comment]
Yeah, I used the latest drivers. In any case, it seems that the flickering affects all AMD cards, while the transparency bug is only an issue on Polaris. (newer architectures weren't tested either so it may or may not happen there)

**Comment by Snakegodeater** [comment]
i will test again with the latest commit and watch all the cutscene all the way up to confirm this !

**Comment by Snakegodeater** [comment]
On v0.0.6-8612  the problem is back !!! Also those strange black lines on that room , Also happening on Plant chapter at strut B room . 
![Χωρίς τίτλο](https://user-images.githubusercontent.com/39376826/63903313-b289a580-ca15-11e9-9947-bceebff31434.jpg)

**Comment by Realmantik** [comment]
Did you try guys run MGS on Linux or old Radeon driver versions?

**Comment by kd-11** [comment]
Use this build to get better RSX captures: https://ci.appveyor.com/project/kd-11/rpcs3/builds/27068332
I tested MGS3 and both AMD and NV had 'transparent' glitches where the depth buffer seems to fall apart for whatever reason. Since it is reported that on MGS2 there is a difference between NV and AMD, I'm more interested in the captures from those scenes.

**Comment by NotBeanz** [comment]
@kd-11 Interestingly the transparency issue that is in-game doesn't show up in the RSX captures with the build that you've provided. Even though Snake has the semi-transparent look when I'm in-game, when I load the capture he shows up fine.
Should I upload it despite all of this?

**Comment by kd-11** [comment]
You should capture with renderdoc if the captures will replay fine (since they will also replay fine on my system anyway).

**Comment by kd-11** [comment]
I forgot to mention: Due to how renderdoc works these days, I'll need two captures, one from polaris (AMD) and one from pascal (NVIDIA). The other alternative would be to capture with OpenGL but this involves editing debug options and is still not guaranteed to work anyway.

**Comment by NotBeanz** [comment]
Okay!
Here's the renderdoc of semi-transparent Snake on the RX 570: https://drive.google.com/open?id=1g3kgVDwPRUB5O9s26qrvEFf3jIHeN2j1

And here's the renderdoc of him looking just fine on the GTX 1050: https://drive.google.com/open?id=13QMoc4yKHHmGfQi8sO_ak6hRxISl01bw

**Comment by kd-11** [comment]
Thank you! This shows the problem much more clearly since I have identical GPUs (RX 470 and 1050Ti)

**Comment by kd-11** [comment]
Ok, it is not AMD that is broken here, it is actually Nvidia, and likely has been broken for decades at this point. The game gives a sample mask with a pattern of 0001000100010001 which clearly skips updating sample 1 of the surface (starting from right, go 0-1-2..upto 15). On nvidia this does not disable sample 1 which raises many questions about how the GPU is even allowed to do that. Given that a PS3 has 4 samples max, I suspect the PS3 does not actually use 1 bit per sample but instead uses 4, so that awful value means to update all samples. I'll need to do some hardware tests, but for now you can just set MSAA level to disabled until I figure out how the PS3 behaves.

**Comment by NotBeanz** [comment]
Thanks for the insight!
Disabling MSAA does indeed fix transparent Snake.

Do you have any idea what could cause the random flickering though?
It affects both MGS2&3, but it's much more prevalent in MGS3. Unfortunately I can't really give more information than the video in my first post. Also, from what I've observed the flicker can sometimes be some text or even the equipped weapon model. This isn't present on NVIDIA either.

**Comment by kd-11** [comment]
Ran some more hw tests, PS3 actually behaves correctly (like AMD) but the pipeline scan was broken on vulkan. This is fixed by https://github.com/RPCS3/rpcs3/pull/6572. Not sure if this negatively affects AMD, I would hope not.

**Comment by micaelis72** [comment]
Just to clarify:

I have a GTX 1060 and I have the same problem.

This is happening on Fedora 30, but this can be a local problem in my system or a general issue for other nvidia users...

**Comment by kd-11** [comment]
The depth issue is different from the sample issue. The samples issue made snake semi-transparent, but not completely. This only happened on AMD.
The depth issue where some stuff is drawn in the wrong order is present on both vendors though. The Z buffer is not initialized which makes me wonder where the data was supposed to come from in the first place. I'll revisit that particular issue in a few days.

**Comment by Snakegodeater** [comment]
Transparent Snake Seems Fixed ! but the square shadows around his model and other models in some areas still remain .

**Comment by kd-11** [comment]
One-line pixels are usually a result of texel snapping/alignment and are usually caused by upscaling. If they're still present at default settings, then provide a new RSX capture with builds from [this PR.](https://github.com/RPCS3/rpcs3/pull/6571)

**Comment by Snakegodeater** [comment]
Ok i'll Report .But i don't know how to do the RSX capture thing . I will try with Anti aliasing off. and then with the default 720p resolution and report back

**Comment by NotBeanz** [comment]
I haven't checked this issue for some time up until today,  and I'm happy to report that the flickers have been fixed with some PR.

Also, as an added bonus the last graphical anomaly in MGS3 (the pilots not being visible in the cockpit) has also been fixed with the latest master.
![screenshot-2019_10_13_22_18_05](https://user-images.githubusercontent.com/54376483/66720612-b166d900-ee07-11e9-879d-b409d6d96726.png)
